node/agent.go

```diff
@@ -262,36 +262,41 @@ func (a *HostAgent) StopService(state *servicestate.ServiceState) error {
 	return ctr.Stop(45 * time.Second)
 }
 
-func reapContainers(maxAge time.Duration) error {
-	containers, err := docker.Containers()
-	if err != nil {
-		return err
-	}
+// reapContainers purges old containers from docker.  It looks at the
+// containers' finish time in order to determine the time of the next purge
+func reapContainers(shutdown <-chan interface{}, minInterval, maxAge time.Duration) {
+	for {
+		var expiration time.Time
 
-	cutoff := time.Now().Add(-maxAge)
-	for _, ctr := range containers {
-		// Do not reap if the container is running, it has never started, or it finished after the cutoff time
-		if ctr.IsRunning() || ctr.State.FinishedAt.Unix() < 0 || ctr.State.FinishedAt.After(cutoff) {
-			continue
+		waitTimeout := maxAge // time to wait for the next container to be purged
+		ctrs, err := docker.Containers()
+		if err != nil {
+			glog.Errorf("Could not look up containers: %s", err)
+			waitTimeout = minInterval
+			goto wait
 		}
 
-		// attempt to delete the container
-		glog.Infof("About to remove container %s", ctr.ID)
-		if err := ctr.Delete(true); err != nil {
-			glog.Errorf("Could not remove container %s: %s", ctr.ID, err)
+		expiration = time.Now().Add(-maxAge)
+		for _, ctr := range ctrs {
+			if finishTime := ctr.State.FinishedAt; finishTime.Unix() <= 0 || ctr.IsRunning() {
+				// container is still running or hasn't started, skip
+				continue
+			} else if timeToLive := expiration.Sub(finishTime); timeToLive <= 0 {
+				// container has exceeded its expiration date
+				if err := ctr.Delete(true); err != nil {
+					glog.Errorf("Could not delete container %s (%s): %s", ctr.Name, ctr.ID, err)
+					waitTimeout = minInterval
+				}
+			} else if timeToLive < waitTimeout {
+				// set the expiration date of the next oldest container not purged
```

Comment:
Is this comment correct? Should it be "Set the time of next purge to the expiration date of oldest unpurged container"?

Also, does this handle the case where a container has been started since this logic was executed and its expiration date falls before timeToLive is hit?